pandas/core/internals/construction.py

```diff
@@ -304,10 +305,26 @@ def ndarray_to_mgr(
     index, columns = _get_axes(
         values.shape[0], values.shape[1], index=index, columns=columns
     )
-    values = values.T
 
     _check_values_indices_shape_match(values, index, columns)
```

Comment:
By moving `values = values.T` below `_check_values_indices_shape_match`, that function doesn't need to take the transposed shapes into account (and can be used for both AM and BM).